fvm/state/state_manager.go

```diff
@@ -0,0 +1,91 @@
+package state
+
+import "fmt"
+
+// StateManager provides active states
+// and facilitates common state management operations
+// in order to make services such as accounts not worry about
+// the state it is recommended that such services wraps
+// an state manager instead of an state itself.
+type StateManager struct {
+	startState  *State
+	activeState *State
+}
+
+// NewStateManager constructs a new state manager
+func NewStateManager(startState *State) *StateManager {
+	return &StateManager{
+		startState:  startState,
+		activeState: startState,
+	}
+}
+
+// State returns the active state
+func (s *StateManager) State() *State {
+	return s.activeState
+}
+
+// StartState returns the start state
+func (s *StateManager) StartState() *State {
+	return s.startState
+}
+
+// MergeStateIntoActiveState allows to merge any given state into the active state
+func (s *StateManager) MergeStateIntoActiveState(other *State) error {
+	return s.activeState.MergeAnyState(other)
+}
+
+// Nest creates a child state and set it as the active state
+func (s *StateManager) Nest() {
+	s.activeState = s.activeState.NewChild()
+}
+
+// RollUp merges the active state into its parent and set the parent as the
+// new active state (if any parents), if merge is set to true, it will merge
+// the delta of the state and if false it ignores the child.
+// if mergeTouches is set it merges the touches to the parent (useful for exec failure cases)
+func (s *StateManager) RollUp(merge bool, mergeTouches bool) error {
```

Comment:
If I understand correctly `RollUp(false, true)` would be used, because the verification nodes still need to confirm this transaction failed.

however `RollUp(true, false)` is not does not seem like it would be ever valid

Reply:
correct, so might actually  break this method into RollUpWithMerge and RollUpOnlyTouch

Reply:
updated at d63488c

Reply:
This description of the method looks extremely complex. Are we 100% sure and certain such a complex logic is the only way to do, what in fact, is quite easy (conceptually) changeset management.
I'm trying to raise it now, because its easier to step back and maybe defer to using `View` structures now instead of maybe refactoring it later